File: php/ConsistentHash/ConsistentHashing.php
<?php
/**
 *
 *
 * Created by PhpStorm
 * User: Alex
 * Date: 2023-07-25 21:11
 */
declare(strict_types=1);

class ConsistentHashing
{

    /**
     * 每个节点对应的虚拟节点个数
     *
     * @var int
     */
    protected $replicas;

    /**
     * 虚拟节点哈希和真实节点值之间的映射关系，key 为虚拟节点的哈希值，value 为节点的标识（如 IP 地址）
     *
     * @var array
     */
    protected $hashMap = [];

    /**
     * 真实节点和虚拟节点之间的映射关系（一个真实节点至少存在一个虚拟节点）
     *
     * @var array
     */
    protected $nodes = [];

    public function __construct(int $replicas)
    {
        if ($replicas <= 0) throw new RuntimeException('replicas must greater than 0');
        $this->replicas = $replicas;
    }

    /**
     * 把字符串转为 32 位无符号整数
     *
     * @param string $str
     * @return int
     */
    public function hash(string $str): int
    {
        // 获取十进制格式的无符号 crc32 校验和的字符串表示形式
        return (int)sprintf('%u', crc32($str));
    }

    public function isEmpty(): bool
    {
        return count($this->hashMap) == 0 || count($this->nodes) == 0;
    }

    /**
     * 添加节点
     *
     * @param string $addr
     * @return void
     */
    public function addNode(string $addr)
    {
        if (! $addr) throw new RuntimeException('the node cannot be null');
        if (isset($this->nodes[$addr])) throw new RuntimeException('the node has already exists');

        for ($i = 0; $i < $this->replicas; $i++) {
            $hash = $this->hash($i . '-' . $addr);
            $this->hashMap[$hash] = $addr;  // 添加虚拟节点和真实节点之间的映射
            $this->nodes[$addr][] = $hash;  // 添加真实节点和虚拟哈希值之间的关系
        }

        // 重新升序排序虚拟节点哈希 key
        ksort($this->hashMap);
    }

    /**
     * 删除节点
     *
     * @param string $addr
     * @return void
     */
    public function removeNode(string $addr)
    {
        if (! isset($this->nodes[$addr])) return;

        // 循环删除虚拟节点
        foreach ($this->nodes[$addr] as $hashVal) {
            unset($this->hashMap[$hashVal]);
        }

        // 删除真实节点
        unset($this->nodes[$addr]);
    }

    /**
     * 根据给定的对象获取最靠近它的那个节点。即为最佳节点。
     *
     * @param string $key
     * @return string
     */
    public function lookup(string $key): string
    {
        if ($this->isEmpty()) {
            throw new RuntimeException('node is empty');
        }

        $hash = $this->hash($key);

        // 因为每添加一个节点时，都会使用 ksort 对虚拟节点哈希做升序排序且每次也都会使用 reset 进行重新指向
        // 因此，这里默认取的就是圆环上最小的一个节点，当作默认值
        $node = current($this->hashMap);

        // 循环获取相近的最优节点：第一个“服务器hash”值大于等于“数据hash”值的就是最优“服务器节点”
        // 因为此时 $this->hashMap 已经是通过虚拟哈希值升序排列，
        // 又因为 php 的 array 遍历均有顺序性，因此这里的遍历顺序也是按照从小到大的顺序进行遍历
        foreach ($this->hashMap as $k => $v) {
            if ($k >= $hash) {
                $node = $v;
                break;
            }
        }

        // 将数组的内部指针指向第一个单元，方便下一次使用 current 时，刚好指向环上最小的一个节点
        reset($this->hashMap);

        return $node;
    }

    public function toArray(): array
    {
        return [
            'replicas' => $this->replicas,
            'hash_map' => $this->hashMap,
            'nodes' => $this->nodes
        ];
    }

}

File: golang/consistentHash/consistentHashing.go
// 一致性 hash 代码参考：https://juejin.cn/post/6871169933150486542
package consistentHash

import (
	"errors"
	"hash/crc32"
	"sort"
	"strconv"
	"sync"
)

// hash 函数需要符合以下要求：
// 1 单调性（唯一） 2平衡性 (数据 目标元素均衡) 3分散性(散列)
type Hash func(data []byte) uint32

type UInt32Slice []uint32

func (s UInt32Slice) Len() int           { return len(s) }
func (s UInt32Slice) Less(i, j int) bool { return s[i] < s[j] }
func (s UInt32Slice) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }

type ConsistentHashBalance struct {
	mux      sync.RWMutex
	hash     Hash                // 哈希函数类型 Hash 的实例，用于计算节点的哈希值
	replicas int                 // 复制因子，表示每个节点在哈希环上的虚拟节点个数
	keys     UInt32Slice         // 已排序的节点哈希切片，用于快速查找节点
	hashMap  map[uint32]string   // 节点哈希和键的映射，键是节点的哈希值，值是节点的标识（如 IP 地址）
	nodes    map[string][]uint32 // 键和节点哈希之间的映射关系，键是节点的标识（如 IP 地址），值是节点的哈希值切片
}

func NewConsistentHashBalance(replicas int, fn Hash) *ConsistentHashBalance {
	if replicas <= 0 {
		panic("replicas must greater than 0")
	}

	m := &ConsistentHashBalance{
		replicas: replicas,
		hash:     fn,
		hashMap:  make(map[uint32]string),
		nodes:    make(map[string][]uint32),
	}
	if m.hash == nil {
		// 最多 32 位，保证是一个 2^32-1 环
		// 默认使用 crc32.ChecksumIEEE 作为哈希函数
		m.hash = crc32.ChecksumIEEE
	}
	return m
}

func (c *ConsistentHashBalance) IsEmpty() bool {
	// 这里只需要知道所有的虚拟环哈希值是否为空即可
	return len(c.keys) == 0
}

// AddNode 方法用来添加节点，参数为节点 key，比如使用 IP 地址
func (c *ConsistentHashBalance) AddNode(addr string) error {
	if "" == addr {
		return errors.New("the node cannot be null")
	}
	if _, ok := c.nodes[addr]; ok {
		return errors.New("the node has already exists")
	}

	c.mux.Lock()
	defer c.mux.Unlock()

	// 结合复制因子计算所有虚拟节点的 hash 值，并存入 m.keys 中，
	// 同时在 m.hashMap 中保存哈希值和 key 的映射
	// 以及在 m.nodes 中保存 key 和哈希值之间的映射关系
	for i := 0; i < c.replicas; i++ {
		hash := c.hash([]byte(strconv.Itoa(i) + "-" + addr))
		c.keys = append(c.keys, hash)               // 记录所有的虚拟节点哈希值
		c.hashMap[hash] = addr                      // 添加虚拟节点和真实节点之间的映射关系
		c.nodes[addr] = append(c.nodes[addr], hash) // 添加真实节点和虚拟哈希值之间的关系
	}

	// 对所有虚拟节点的哈希值进行升序排序，方便之后进行二分查找
	// 这里和 php 处理有点儿区别，php 可以直接对关联数组进行排序，比较感叹 php 的数组功能是真的强大，😄
	sort.Sort(c.keys)

	return nil
}

// RemoveNode 方法用于移除缓存节点，参数为节点key，比如使用IP
func (c *ConsistentHashBalance) RemoveNode(addr string) {
	c.mux.Lock()
	defer c.mux.Unlock()

	if _, ok := c.nodes[addr]; !ok {
		// 根本就不存在改节点时，则不需要被移除
		return
	}

	// 需要移除的节点的哈希值
	removeHashes := c.nodes[addr]

	// 从 keys 和 hashMap 中移除对应哈希值的节点
	newKeys := make(UInt32Slice, 0, len(c.keys))
	for _, key := range c.keys {
		if !contains(removeHashes, key) {
			newKeys = append(newKeys, key)
		} else {
			// 需要一个一个的删除虚拟节点和真实节点之间的映射关系
			delete(c.hashMap, key)
		}
	}
	c.keys = newKeys

	// 删除掉真实节点和虚拟节点之间的映射关系
	delete(c.nodes, addr)
}

// contains 函数用于判断切片中是否包含某个元素
func contains(s []uint32, val uint32) bool {
	for _, v := range s {
		if v == val {
			return true
		}
	}
	return false
}

// Lookup 方法根据给定的对象获取最靠近它的那个节点
func (c *ConsistentHashBalance) Lookup(key string) (string, error) {
	if c.IsEmpty() {
		return "", errors.New("node is empty")
	}

	hash := c.hash([]byte(key))

	// 通过二分查找函数获取最优节点，第一个"服务器hash"值大于等于"数据hash"值的就是最优"服务器节点"
	idx := sort.Search(len(c.keys), func(i int) bool { return c.keys[i] >= hash })

	// 如果查找结果 大于 服务器节点哈希数组的最大索引，表示此时该对象哈希值位于最后一个节点之后，那么放入第一个节点中
	if idx == len(c.keys) {
		idx = 0
	}
	c.mux.RLock()
	defer c.mux.RUnlock()
	return c.hashMap[c.keys[idx]], nil
}
